### ISSUE TITLE
Feature request: precommand support (e.g. sudo)

### ISSUE THREAD

**Issue by Willy-JL** [chrisant996/clink#261]
Ideal behavior would be:
 - detect precommand (I only mentioned sudo because I can't think of anything else like this on Windows)
 - optionally highlight it (`color.precommand`)
 - continue with `exec` highlight and normal argmatching, suggestions and completions

I tried to implement this myself but C and C++ are beyond me and I could only get it to ignore precommands and continue proper exec highlighting. Couldn't manage to shift the argmatching, suggestions and completions to after the precommand. Nonetheless here is what my failed attempt looked like, in case it is of any use to implement this feature.

```diff
diff --git a/clink/lua/scripts/arguments.lua b/clink/lua/scripts/arguments.lua
index a4a7fffa..54f2edc0 100644
--- a/clink/lua/scripts/arguments.lua
+++ b/clink/lua/scripts/arguments.lua
@@ -1844,11 +1844,19 @@ function argmatcher_classifier:classify(commands)
         local line_state = command.line_state
         local word_classifier = command.classifications

-        local argmatcher, has_argmatcher, extra_words = _find_argmatcher(line_state, true)
-        local command_word_index = line_state:getcommandwordindex()
-
         local word_count = line_state:getwordcount()
+        local command_word_index = line_state:getcommandwordindex()
         local command_word = line_state:getword(command_word_index) or ""
+
+        for _,precommand in ipairs({"sudo"}) do
+            if command_word == precommand then
+                command_word_index = command_word_index + 1
+                command_word = line_state:getword(command_word_index) or ""
+            end
+        end
+
+        local argmatcher, has_argmatcher, extra_words = _find_argmatcher(line_state, true)
+
         if #command_word > 0 then
             local info = line_state:getwordinfo(command_word_index)
             local m = has_argmatcher and "m" or ""
```

I can tell that this would be a great undertaking and would likely take a lot of reworking, so I'm happy if this is even just considered.

On a side note, thank you so much for maintaining this project. I just discovered clink the other day from a friend and I already can't live without it.

**Comment by pukkandan** [comment]
> I only mentioned sudo because I can't think of anything else like this on Windows

`cmd /c` would be an example

**Comment by Willy-JL** [comment]
> `cmd /c` would be an example

True, but that would get even more complicated since they would count as two separate words and so two separate commands or whatever (just saying this based on what I saw in the code, but by all means this is a good example)

**Comment by chrisant996** [comment]
Just last week, I started thinking about a similar case I have:

```
i {dir} {command line}
```

Which sets cwd to {dir} and runs {command line} and restores the original cwd.

This can be sort of done already with the new delayinit argmatchers ("Adaptive Argmatchers" in the docs) plus the new `clink.getargmatcher()`.  But it isn't possible to do a complete job of it yet, without additional changes to Clink.

I think the way to do it would be to have a variation on `:addarg()` that makes the rest of the input line get treated as though `&` had been encountered, i.e. start a new command parse.

That would be relatively straightforward to add, and would offer very high flexibility and power.  It's a little trickier to get the input line coloring to apply `color.argmatcher` in the "embedded command", and to figure out how to deal with doskey aliases in the embedded command.

**Comment by chrisant996** [comment]
This will be available in v1.3.13.

The sudo example is simple:
```lua
clink.argmatcher('sudo'):chaincommand()
```

You may extend the argmatcher with flags, etc.

And regarding the `i {dir} {command}` script I mentioned earlier ... Clink is now able to natively handle that, instead of needing a .cmd script.  The benefit of Clink providing the `i` command natively via Lua is that the command line can be handled exactly as intended by the user.  Using a .cmd script runs afoul of CMD batch argument parsing quirks if the command line contains commas or equal signs or various other characters.

After v1.3.13 is published, I will post the `i.lua` script in the Discussion section.
